src/inmanta/deploy/work.py

```diff
@@ -0,0 +1,424 @@
+
+    def __lt__(self, other: object) -> bool:
+        if not isinstance(other, TaskQueueItem):
+            return NotImplemented
+        return (self.task.priority, self.insert_order) < (other.task.priority, other.insert_order)
+
+
+class AgentQueues(Mapping[Task, PrioritizedTask[Task]]):
+    """
+    Per-agent priority queue for ready-to-execute tasks. Clients must not interact with the underlying priority queues directly,
+    only through the queue manipulation methods offered by this class.
+
+    Each Task will be queued at most once. If an identical Task is queued later on, it will not be queued separately. It may
+    however affect the current task's priority, causing it to be executed earlier than it would have otherwise.
+    """
+
+    # TODO(ticket): relies on undocumented asyncio.PriorityQueue._queue field and the fact that it's a heapq, can we do something about that?
+
+    def __init__(self) -> None:
+        self._agent_queues: dict[str, asyncio.PriorityQueue[TaskQueueItem]] = {}
+        # can not drop tasks from queue without breaking the heap invariant, or potentially breaking asyncio.Queue invariants
+        # => take approach suggested in heapq docs: simply mark as deleted.
+        # => Keep view on all active tasks for a given resource, which also doubles as lookup for client operations
+        # Only tasks in this collection are considered queued, as far as this class' client interface is concerned.
+        self._tasks_by_resource: dict[ResourceIdStr, dict[Task, TaskQueueItem]] = {}
+        # monotonically rising value for item insert order
+        # use simple counter rather than time.monotonic_ns() for performance reasons
+        self._entry_count: int = 0
+
+    def get_tasks_for_resource(self, resource: ResourceIdStr) -> set[Task]:
+        """
+        Returns all queued tasks for a given resource id.
+        """
+        return set(self._tasks_by_resource.get(resource, {}).keys())
+
+    def remove(self, task: Task) -> int:
+        """
+        Removes the given task from its associated agent queue. Raises KeyError if it is not in the queue.
+        Returns the priority at which the deleted task was queued.
+        """
+        tasks: dict[Task, TaskQueueItem] = self._tasks_by_resource.get(task.resource, {})
+        queue_item: TaskQueueItem = tasks[task]
+        del tasks[task]
+        if not tasks:
+            del self._tasks_by_resource[task.resource]
+        return queue_item.task.priority
+
+    def discard(self, task: Task) -> Optional[int]:
+        """
+        Removes the given task from its associated agent queue if it is present.
+        Returns the priority at which the deleted task was queued, if it was at all.
+        """
+        try:
+            return self.remove(task)
+        except KeyError:
+            return None
+
+    def _queue_item_for_task(self, task: Task) -> Optional[TaskQueueItem]:
+        return self._tasks_by_resource.get(task.resource, {}).get(task, None)
```

Comment:
Why not make Tasks the same as TaskQueueItems or carry a reference to them?

Reply:
As I see it, `Task` is the high level declaration that such a resource action should be performed. By containing it to only the declaration of this intent, it can be used for comparison, e.g. to check whether a certain task is already scheduled. Perhaps `Task` is not the best name for it.

That said, I'll consider the value of keeping it like that versus what you suggest. I do think that there is one easy win in the direction you suggest in merging `PrioritizedTask` and `TaskQueueItem` into one. I think they had a separate value in my initial implementation but that ceased to be, so it would be cleaner to merge them. Then we'd only have two levels: the "intent" and the "task".

Reply:
`Task` also allows the queue consumer to work with a simpler interface, i.e. just the intent.